src/core/unit.h

```diff
@@ -125,11 +135,13 @@ typedef struct Unit {
 
         Set *aliases; /* All the other names. */
 
-        /* For each dependency type we maintain a Hashmap whose key is the Unit* object, and the value encodes why the
-         * dependency exists, using the UnitDependencyInfo type */
-        Hashmap *dependencies[_UNIT_DEPENDENCY_MAX];
+        /* For each dependency type we can look up another Hashmap with this, whose key is a Unit* object,
+         * and whose value encodes why the dependency exists, using the UnitDependencyInfo type. i.e. a
+         * Hashmap(UnitDependency → Hashmap(Unit* → UnitDependencyInfo)) */
+        Hashmap *dependencies;
```

Comment:
How does this affect memory usage in the typical scenario?

Reply:
I didn't measure it (it's not obvious to measure, due to the allocation cache). I'd expect moderate savings, since we no longer allocate space for per-dependency-type pointers to hashmap we almost never use. Given that sparely populated hashmaps are a single cache-backed allocations typically I think we basically make the Unit structure allocation quite a bit shorter with the Hashmap taking up much less space.

or to say this in math: we have now 33 dep types. on 64bit an array with a pointer for each is 264 bytes. A sparsely populated Hashmap costs only 48 bytes however (and as we have an allocation cache only one extra ptr for the cache maintainance, plus one pointer to the Hashmap from Unit). Hence 64 bytes with this PR vs. 264 bytes without it. Runtime behaviour should be slightly better with the PR as when iterating through the deps we only need to iterate through the dep types we actually have deps for instead of all dep types all the times.

(That is, if I didn#t make a mistake in my calculations)